Docs: Custom Types & Actions (#467)
amplication/docs: docs/premium-features/custom-types-and-actions.md

To disable an existing DTO, click on the toggle and move it into the disabled state. The dot in the left sidebar will now change colors into Red. You won't be able to use this DTO within the module.

To delete a DTO entirely, click on the delete button next to the toggle. A confirmation popup will appear. Click on the delete button again to fully delete the DTO.

:::note
Be cautious when deleting DTOs, as it may impact other parts of your application that depend on them.
:::

![Edit, Disable, and Delete DTO](./assets/custom-types-and-actions/edit-disable-delete-dto.png)

## Actions

Actions represent the API endpoints that define the operations available in your application. They specify the HTTP methods, paths, input, and output data structures for each endpoint.

Actions use DTOs to define their input and output data. Each Action can have a request DTO (for input) and a response DTO (for output). Amplication generates built-in Actions based on your application's entities, but you can also create custom Actions to implement specific functionality.

### Create Action

To create a new Action, follow these steps:

> 1. Click on the "Add Action" button in the top right of the APIs tab.

Reviewer: Still missing it ^^^

2. Provide a name for your new Action.
3. Click the "Create Action" button to finalize the creation process.

![Edit, Disable, and Delete DTO](./assets/custom-types-and-actions/create-action.png)

### Enable and Disable Actions

Amplication gives you a complete overview of all Actions associated with a specific Module.

This is useful when you want to temporarily disable certain endpoints without completely removing them from your application.

To enable or disable an Action, simply toggle the switch next to the Action's name in the APIs tab.

![Enable and Disable Actions](./assets/custom-types-and-actions/all-actions.png)

### Edit and Delete Action

Similar to DTOs, you can edit, delete, move, and duplicate Actions within the APIs tab. When editing an Action, you can modify its HTTP method, path, input DTO, output DTO, and other settings as needed. Remember to save your changes when you're done.

> 

Reviewer: Same as the DTO comment- I think we need to better explain how to add / edit the Action itself- choosing input and output DTOs, defining the REST and GQL parameters, including the "Input Source" for REST (see below) etc. 
![image](https://github.com/amplication/docs/assets/149934977/4a4aefad-75d0-4799-bade-43f52b217088)

![Edit Actions](./assets/custom-types-and-actions/edit-action.png)

## Conclusion

> Custom Types & Actions in Amplication provide a powerful and flexible way to define and manage your application's data structures and API endpoints. By leveraging custom DTOs and Actions, you can create a tailored and efficient API that meets your specific requirements.

Reviewer: ... and generate the production-ready code for it in your repo etc.

:::

## What Are Types And Actions?

In Amplication, Types and Actions are fundamental concepts that define the data structures and API endpoints of your application.

### Types

Types in Amplication refer to the data structures used to define the shape and format of your application's data. There are two main categories of types:

1. **Data Transfer Objects (DTOs)**: DTOs are custom data structures that define the input and output data for your API endpoints (Actions). They encapsulate the data that is transferred between the client and the server or between different services within your application.

2. **Primitives**: Primitives are the basic data types supported by Amplication, such as String, Integer, Boolean, Date, and JSON. These primitives can be used as fields within your DTOs or as standalone types for your Actions.

### Actions

Actions in Amplication represent the API endpoints that define the operations available in your application. Each Action specifies the input, output, and other configurations specific to the API type (REST or GraphQL).

For [REST API](/api/#rest-apis), an Action defines:

> - HTTP method (GET, POST, PUT, DELETE, etc.)

Reviewer 2: ```suggestion
- HTTP verb (GET, POST, PUT, DELETE, etc.)
```

> - Path (e.g., `/users` or `/products/:id`)

Reviewer 2: ```suggestion
- Route (e.g., `/users` or `/products/:id`)
```

- Input parameters (query, path, or request body)
- Output data structure (response body)

For [GraphQL API](/api/#graphql-apis), an Action defines:

- Operation type (Query or Mutation)
- Input type (arguments)
- Output type (return type)

Actions can be of two types:

1. **Default CRUD Actions**: Amplication automatically generates default CRUD (Create, Read, Update, Delete) Actions based on your application's entities. These Actions provide standard API endpoints for interacting with your data.

> 

Reviewer 2: Consider if we want to add information about the default *relation* CRUD operations (e.g. get all books for author, get author for book, connect/disconnect book to author) 
@PazYanoverr - it's not mandatory, for your consideration

Reviewer: Sure, @dericksozo - Let's add it.

     - `entities` (Read all)
     - `entity(id: ID!)` (Read by ID)
     - `createEntity(input: CreateEntityInput!)` (Create)
     - `updateEntity(id: ID!, input: UpdateEntityInput!)` (Update)
     - `deleteEntity(id: ID!)` (Delete)

2. **Custom Actions**: Custom Actions are user-defined API endpoints that extend the functionality of your application beyond the default CRUD operations. You can create Custom Actions to implement specific business logic, perform complex data transformations, or integrate with external services.

## Accessing Custom Types & Actions

To access the Custom Types & Actions feature, navigate to the new **APIs** tab in your Amplication dashboard.

The APIs tab provides a centralized view of all your project's Modules, DTOs, and Actions. It's also further divided into sections for each Module, making it easy to browse and manage your project's custom types and actions.

![APIs Tab Overview](./assets/custom-types-and-actions/api-tab-overview.png)

## Modules

Modules are organizational units that group related DTOs and Actions together. They help structure your application in a logical and maintainable way.

> Modules are closely tied to entities in your application. Each entity is generated as a Module containing all the DTO and API code related to that Module. You can also create additional Modules based on your business needs and requirements.

Reviewer 2: I would rephrase "modules are closely tied to entities" to ensure people understand better it's not mandatory.
Perhaps something like - "A module is encapsulating a unit of business logic that is related to one application domain. By default, Amplication creates a module for each DB entity with all the CRUD operations, and you can add your own custom modules based on your business needs and requirements. You can add custom DTOs and custom Actions to a module - whether it's an entity-related module, or a custom module you've created"

### Create DTO

To create a new DTO, follow these steps:

:::note
If you create a new DTO / Enum from the "All Modules" screen, you can choose which Module it will reside in.
:::

#### DTO Name and Description

1. Click on the "Add DTO" button in the top right of the APIs tab.
2. In the dropdown, click on "Add DTO".
3. Provide a name for your DTO, such as "MyNewDtoName". This will automatically populate the Name field with the provided name.
4. Optionally, add a description for your DTO.

#### Adding Properties

1. In the "Properties" section, click on "Add property" to define the DTO's properties.
2. For each property:
   - Specify the Name of the property (e.g., "firstProperty").
   - Choose the Type of the property from the dropdown (e.g., "String", "Number", "Boolean", "Date", "JSON", "DTO", "Enum").
   - If needed, you can change the property to an array by toggling the "Array" switch.

>    - Mark the property as optional by checking the "Optional" checkbox.

Reviewer 2: ```suggestion
   - If needed, mark the property as optional by checking the "Optional" checkbox.
```

3. Repeat step 2 for each additional property you want to add to the DTO.

### Create Enum

To create a new Enum, follow these steps:

#### Enum Name and Description

1. Click on the "Add DTO" button in the top right of the APIs tab.
2. In the dropdown, click on "Add Enum".
3. Provide a name for your Enum, such as "MyNewEnum". This will automatically populate the Name field with the provided name.
4. Optionally, add a description for your Enum.

#### Adding Members

1. In the "Members" section, click on "Add member" to define the Enum's members.
2. For each member:
   - Specify the Name of the member.
   - Provide the Value for the member.
3. Repeat step 2 for each additional member you want to add to the Enum.

### Edit DTO or Enum

To edit an existing DTO, click on the DTO within its associated Module in the APIs tab. You can modify the DTO's name, description, and properties as needed.

## Actions

> Actions represent the API endpoints that define the operations available in your application. They specify the HTTP methods, paths, input, and output data structures for each endpoint.

Reviewer 2: This text is more REST-centric than GQL centric. Not sure it's a problem, but a point to consider.
Anyway, if we leave it as is, let's change methods to verbs and paths to routes

Actions use DTOs to define their input and output data. Each Action can have a request DTO (for input) and a response DTO (for output). Amplication generates built-in Actions based on your application's entities for CRUD operations, but you can also create custom Actions to implement specific functionality.

:::note

> The generated Amplication service has support for HTTP mode only. Our [HTTPs plugin](https://amplication.com/plugins/transport-https) adds support for HTTPS mode and TLS termination at the app level.

Reviewer 2: ```suggestion
By default, the generated Amplication service exposes its endpoints over HTTP to allow HTTPS termination at a higher level (like an API Gateway, etc.). However, our [HTTPs plugin](https://amplication.com/plugins/transport-https) adds support for HTTPS mode and TLS termination at the app level.
```

:::

### Create Action

To create a new Action, follow these steps:

#### Module Name and Description

1. Click on the "Add Action" button in the top right of the APIs tab.
2. Provide a Display Name for your new Action, such as "My New Module Name". This will automatically populate the Name field with a generated name like "MyNewModuleName".
3. Optionally, add a description for your Action.

:::note
If you create a new Action from the "All Modules" screen, you can choose which Module the Action will reside in.
:::

#### Choosing Types

1. In the "Types" section, select the input and output types for your Action:
   - For the Input Type, choose the type of the property from the dropdown (e.g., "String", "Number", "Boolean", "Date", "JSON", "DTO", "Enum").
   - If you choose "DTO", click "Select..." to pick an existing DTO or create a new one.
   - Repeat the same steps for the Output Type.
2. If needed, you can change the Input Type or Output Type to an array by toggling the "Array" switch.

#### REST API Settings

> 1. Choose the appropriate HTTP Method from the dropdown (e.g., "get", "post", "put", "delete").

Reviewer 2: ```suggestion
1. Choose the appropriate HTTP Verb from the dropdown (e.g., "get", "post", "put", "delete").
```

> 2. Specify the Path for your Action (e.g., "/id/my-new-module-name").

Reviewer 2: ```suggestion
2. Specify the Route for your Action (e.g., "/id/my-new-module-name").
```

3. Select the Input Source for your Action's parameters (e.g., "body", "query", "params", or "split") from the dropdown.

#### GraphQL API Settings

Choose the Operation Type for your Action (e.g., "query" or "mutation").

![Create Action](./assets/custom-types-and-actions/create-action.png)

### Enable and Disable Actions

Amplication gives you a complete overview of all Actions associated with a specific Module.

This is useful when you want to temporarily disable certain endpoints without completely removing them from your application.

To enable or disable an Action, simply toggle the switch next to the Action's name in the APIs tab.

:::caution
Disabling an Action will cause Amplication not to generate the code for that Action.
:::

![Enable and Disable Actions](./assets/custom-types-and-actions/all-actions.png)

### Edit and Delete Action

> To edit an Action, click on the Action within its associated Module in the APIs tab. You can modify the Action's properties, such as the HTTP method, path, input DTO, output DTO, and other settings as needed.

Reviewer 2: ```suggestion
To edit an Action, click on the Action within its associated Module in the APIs tab. You can modify the Action's properties, such as the HTTP verb, route, input DTO, output DTO, and other settings as needed.
```

To delete an Action, click on the delete button next to the Action's name. Be cautious when deleting Actions, as it may impact other parts of your application that depend on them.

:::note
Actions generated by Amplication **cannot be deleted**.
:::

## Conclusion

> Custom Types & Actions in Amplication provide a powerful and flexible way to define and manage your application's data structures and API endpoints. By leveraging custom DTOs and Actions, you can create a tailored and efficient API that meets your specific requirements and generates the production-ready code for it in your repository.

Reviewer 2: @PazYanoverr, perhaps elaborate here on benefits like a single source of truth for APIs and DTOs, fast creation of all your APIs, and standardization.

Reviewer: Agree, great points!
@dericksozo - Let's emphasize it not only at the conclusion but also at the beginning of the doc.

# Custom Types & Actions

Custom Types & Actions is an enhanced feature that gives you full control over your Data Transfer Objects (DTOs) and APIs. It allows you to automatically generate high-quality, efficient code for both the default CRUD operations and your custom APIs that utilize custom DTOs, all within your REST and/or GraphQL APIs.

Custom Types & Actions let you define custom data structures, create API endpoints, and manage the relationships between them, all within the Amplication UI. These changes are automatically translated into production-ready code in your repository, ensuring the best quality and ,efficiency.

:::note
The Custom Types & Actions feature is available for Amplication's **Enterprise** plan, including trial plans.
:::

> 

Reviewer: I think we need to emphasize and clarify somewhere at the beginning of the doc, that Actions = APIs, and Types = DTOs, and primitives like String, Integer etc.
I noticed you mentioned it regarding the Actions, but only later in the doc, and I believe it should be somewhere at its beginning.

## Accessing Custom Types & Actions

To access the Custom Types & Actions feature, navigate to the new **APIs** tab in your Amplication dashboard.

The APIs tab provides a centralized view of all your project's Modules, DTOs, and Actions. It's also further divided into sections for each Module, making it easy to browse and manage your project's custom types and actions.

![APIs Tab Overview](./assets/custom-types-and-actions/api-tab-overview.png)

## Modules

Modules are organizational units that group related DTOs and Actions together. They help structure your application in a logical and maintainable way.

> Modules are closely tied to entities in your application, with each entity being generated as a Module containing all the entity's code (DTOs, APIs, and more). Each Module is generated as a folder containing all the DTOs and APIs code related to that Module. You can also create additional Modules based on your business needs and requirements.

Reviewer: It seems to me like we have some kind of redundant here- "each entity being generated as a Module containing all the entity's code (DTOs, APIs, and more)" and "Each Module is generated as a folder containing all the DTOs and APIs code related to that Module". Can we unite it into a single clearer sentence?

### Create DTO

To create a new DTO, follow these steps:

#### DTO Name and Description

1. Click on the "Add DTO" button in the top right of the APIs tab.
2. In the dropdown, click on "Add DTO".
3. Provide a name for your DTO, such as "MyNewDtoName". This will automatically populate the Name field with the provided name.
4. Optionally, add a description for your DTO.

#### Adding Properties

1. In the "Properties" section, click on "Add property" to define the DTO's properties.
2. For each property:
   - Specify the Name of the property (e.g., "firstProperty").
   - Choose the Type of the property from the dropdown (e.g., "String", "Number", "Boolean", "Date", "JSON", "DTO", "Enum").
   - If needed, you can change the property to an array by toggling the "Array" switch.
   - Mark the property as optional by checking the "Optional" checkbox.
3. Repeat step 2 for each additional property you want to add to the DTO.

> 4. If you need to define a union type for a property, click on "+ Add union type" and specify the types for the union.

Reviewer: Let's remove the union description- as we will soon delete this code and avoid releasing this option (at this dev phase)

#### Finalizing DTO Creation

> Click the "Create DTO" button to finalize the creation process.

Reviewer: Not sure which "Create DTO" you refer to here. After the DTO is created, adding properties is saved automatically, and not need to press any button to apply it

After completing these steps, your new DTO will be created within the selected Module, and you can further customize its properties and relationships as needed.

### Create Enum

To create a new Enum, follow these steps:

#### Enum Name and Description

1. Click on the "Add DTO" button in the top right of the APIs tab.
2. In the dropdown, click on "Add Enum".
3. Provide a name for your Enum, such as "MyNewEnum". This will automatically populate the Name field with the provided name.
4. Optionally, add a description for your Enum.

#### Adding Members

1. In the "Members" section, click on "Add member" to define the Enum's members.
2. For each member:
   - Specify the Name of the member.
   - Provide the Value for the member.
3. Repeat step 2 for each additional member you want to add to the Enum.

#### Finalizing Enum Creation

> Click the "Create Enum" button to finalize the creation process.

Reviewer: Same as previous message- there is no "Create Enum" button to finalize the creation / edition. Is it saved automatically

After completing these steps, your new Enum will be created within the selected Module, and you can use it as a type for properties in your DTOs or as input/output types for your Actions.

### Edit DTO or Enum

> To edit an existing DTO, click on the DTO within its associated Module in the APIs tab. You can modify the DTO's name, description, and properties as needed. Remember to save your changes when you're done.

Reviewer: "Remember to save your changes when you're done." - no such option / need

Actions represent the API endpoints that define the operations available in your application. They specify the HTTP methods, paths, input, and output data structures for each endpoint.

Actions use DTOs to define their input and output data. Each Action can have a request DTO (for input) and a response DTO (for output). Amplication generates built-in Actions based on your application's entities for CRUD operations, but you can also create custom Actions to implement specific functionality.

:::note
The generated Amplication service has support for HTTP mode only. Our [HTTPs plugin](https://amplication.com/plugins/transport-https) adds support for HTTPS mode and TLS termination at the app level.
:::

### Create Action

To create a new Action, follow these steps:

#### Module Name and Description

1. Click on the "Add Action" button in the top right of the APIs tab.
2. Provide a Display Name for your new Action, such as "My New Module Name". This will automatically populate the Name field with a generated name like "MyNewModuleName".
3. Optionally, add a description for your Action.

#### Choosing Types

1. In the "Types" section, select the input and output types for your Action:

>    - For the Input Type, choose "DTO" from the dropdown and click "Select..." to pick an existing DTO or create a new one.

Reviewer: Let's add the the Input and Output type can be also primitive, not only DTO, and it can be chosen from the drop-down list

   - For the Output Type, choose "DTO" from the dropdown and click "Select..." to pick an existing DTO or create a new one.
2. If needed, you can change the Input Type or Output Type to an array by toggling the "Array" switch.

#### REST API Settings

1. Choose the appropriate HTTP Method from the dropdown (e.g., "get", "post", "put", "delete").
2. Specify the Path for your Action (e.g., "/id/my-new-module-name").
3. Select the Input Source for your Action's parameters (e.g., "body", "query", "params", or "split") from the dropdown.

#### GraphQL API Settings

Choose the Operation Type for your Action (e.g., "query" or "mutation").

#### Finalizing Action Creation

> Click the "Create Action" button to finalize the creation process.

Reviewer: No such button.... same comment

To enable or disable an Action, simply toggle the switch next to the Action's name in the APIs tab.

:::caution
Disabling an Action will cause Amplication not to generate the code for that Action.
:::

![Enable and Disable Actions](./assets/custom-types-and-actions/all-actions.png)

### Edit and Delete Action

To edit an Action, click on the Action within its associated Module in the APIs tab. You can modify the Action's properties, such as the HTTP method, path, input DTO, output DTO, and other settings as needed. Remember to save your changes when you're done.

> To delete an Action, click on the delete button next to the Action's name. Be cautious when deleting Actions, as it may impact other parts of your application that depend on them.

Reviewer: Let's mention here that Amplication Actions (built-in, or such) cannot be deleted

:::note

> Moving or duplicating Actions (and DTOs) is currently not be supported.

Reviewer: Any reason to mention something that we don't have / support?
